patches/0058-libavocdec-qsvenc_hevc-encode-RGB-format-rawvideo.patch

```diff
@@ -71,27 +71,23 @@ index d26e8115f0..661a9d7fa0 100644
      switch (q->param.mfx.RateControlMethod) {
      case MFX_RATECONTROL_CBR:
      case MFX_RATECONTROL_VBR:
-@@ -1596,8 +1603,16 @@ static int submit_frame(QSVEncContext *q, const AVFrame *frame,
-             qf->surface.Info.PicStruct |= MFX_PICSTRUCT_FRAME_TRIPLING;
+@@ -1609,6 +1616,14 @@ static int submit_frame(QSVEncContext *q, const AVFrame *frame,
+             qf->surface.Data.V     = qf->surface.Data.UV + 2;
+             break;
  
-         qf->surface.Data.PitchLow  = qf->frame->linesize[0];
--        qf->surface.Data.Y         = qf->frame->data[0];
--        qf->surface.Data.UV        = qf->frame->data[1];
-+        if (frame->format == AV_PIX_FMT_X2RGB10LE ||
-+            frame->format == AV_PIX_FMT_BGRA) {
++        case AV_PIX_FMT_X2RGB10LE:
++        case AV_PIX_FMT_BGRA:
 +            qf->surface.Data.B         = qf->frame->data[0];
 +            qf->surface.Data.G         = qf->frame->data[0] + 1;
 +            qf->surface.Data.R         = qf->frame->data[0] + 2;
 +            qf->surface.Data.A         = qf->frame->data[0] + 3;
-+        } else {
-+            qf->surface.Data.Y         = qf->frame->data[0];
-+            qf->surface.Data.UV        = qf->frame->data[1];
-+        }
-     }
- 
-     qf->surface.Data.TimeStamp = av_rescale_q(frame->pts, q->avctx->time_base, (AVRational){1, 90000});
++            break;
++
+         default:
+             /* should not reach here */
+             av_assert0(0);
```

Comment:
With this patch applied, new code looks like:

```c
        qf->surface.Data.PitchLow  = qf->frame->linesize[0];
        qf->surface.Data.Y         = qf->frame->data[0];
        qf->surface.Data.UV        = qf->frame->data[1];

        /* The SDK checks Data.V when using system memory for VP9 encoding */
        switch (frame->format) {
        case AV_PIX_FMT_NV12:
            qf->surface.Data.V     = qf->surface.Data.UV + 1;
            break;

        case AV_PIX_FMT_P010:
            qf->surface.Data.V     = qf->surface.Data.UV + 2;
            break;

        case AV_PIX_FMT_X2RGB10LE:
        case AV_PIX_FMT_BGRA:
            qf->surface.Data.B         = qf->frame->data[0];
            qf->surface.Data.G         = qf->frame->data[0] + 1;
            qf->surface.Data.R         = qf->frame->data[0] + 2;
            qf->surface.Data.A         = qf->frame->data[0] + 3;
            break;

        default:
            /* should not reach here */
            av_assert0(0);
            break;
        }
```

I'm not familiar with the ffmpeg code and structures... but now `Data.Y` and `Data.UV` are set unconditionally (as opposed to the original patch).  For RGB format, will unconditional access to `qf->frame->data[1]` to set `Data.UV` always succeed and not crash?

Reply:
[https://github.com/Intel-Media-SDK/MediaSDK/blob/4a39890fc46c09726968e1c75a7f567dd7426b13/api/include/mfxstructures.h#L237](https://github.com/Intel-Media-SDK/MediaSDK/blob/4a39890fc46c09726968e1c75a7f567dd7426b13/api/include/mfxstructures.h#L237) Because UV and G are in the same union, UV can be overwrite when encode RGB.

Reply:
Right.  But, my concern is about access to `qf->frame->data[1]`?  How is this data defined?

Reply:
This data is defined as a array of pointer:
`uint8_t *data[8];`

Reply:
👍